### DISCUSSION TITLE
InsertOrUpdate with multiple rows

### DISCUSSION THREAD

**Discussion by ademchenko** [linq2db/linq2db#3655] [Q&A]
Hello, as far as I understand, InsertOrUpdate method in linq2db is implemented for Postgres as `INSERT INTO ... VALUES (...) (...) (...) ON CONFLICT...`
The thing is that SQL clause does support multiple inserting values but linq2db does not. Is there a similar method in linq2db with the multiple inserts and if not, for what technical reason?

**Comment by to11mtm** [comment]
@ademchenko 

Linq2db support multiple inserts (via BulkCopy with `BulkCopyType.MultipleRows`) but not with an `ON CONFLICT` at this time.

I cannot speak with full certainty but there are probably a few factors:

 - LinqToDb strives for broad cross compatibility over the baseline API. Most other DBs implement similar(ish) behavior via MERGE, which has a lot of work around but there is no 'translation' from Merge to Postgres ON CONFLICT at this time.
 
 - Postgres should be getting merge support, and there is an issue to track [adding support to LinqToDb](https://github.com/linq2db/linq2db/issues/3584)

That said, there may be some opportunity for adding more flexibility to the Bulk Copy API; at one point I looked into adding 'returning' support and saw there was opportunity for a lot of flexibility with some refactoring.
